core/src/main/java/org/mule/runtime/core/internal/type/catalog/DefaultArtifactTypeLoader.java

```diff
@@ -0,0 +1,134 @@
+/*
+ * Copyright (c) MuleSoft, Inc.  All rights reserved.  http://www.mulesoft.com
+ * The software in this package is published under the terms of the CPAL v1.0
+ * license, a copy of which has been included with this distribution in the
+ * LICENSE.txt file.
+ */
+package org.mule.runtime.core.internal.type.catalog;
+
+import static java.lang.String.format;
+import static java.util.Optional.ofNullable;
+import static java.util.stream.Collectors.joining;
+import static org.mule.metadata.api.utils.MetadataTypeUtils.getTypeId;
+import static org.mule.runtime.extension.api.util.ExtensionMetadataTypeUtils.getAlias;
+
+import org.mule.metadata.api.TypeLoader;
+import org.mule.metadata.api.model.MetadataType;
+import org.mule.metadata.api.model.ObjectType;
+import org.mule.metadata.catalog.api.PrimitiveTypesTypeLoader;
+import org.mule.runtime.api.lifecycle.Initialisable;
+import org.mule.runtime.api.lifecycle.InitialisationException;
+import org.mule.runtime.api.meta.model.ExtensionModel;
+import org.mule.runtime.api.metadata.ArtifactTypeLoader;
+import org.mule.runtime.core.api.extension.ExtensionManager;
+
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.inject.Inject;
+
+/**
+ * a {@link ArtifactTypeLoader} for obtaining types available in the context of the current artifact. It accepts primitive type
+ * names (such as string, number, etc.) and can also access types defined in other extensions by using the
+ * {@code <extension_namespace>:<type_name>} syntax.
+ *
+ * @since 4.5.0
+ */
+public class DefaultArtifactTypeLoader implements ArtifactTypeLoader, Initialisable {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(DefaultArtifactTypeLoader.class);
+
+  private final TypeLoader primitivesTypeLoader = new PrimitiveTypesTypeLoader();
+  private final TypeLoader specialTypesLoader = new SpecialTypesTypeLoader();
+
+  private Map<String, Collection<ObjectType>> typesByExtension;
+  private Map<String, Optional<MetadataType>> loadedTypes;
+
+  @Inject
+  private ExtensionManager extensionManager;
+
+  private Collection<ExtensionModel> extensionModels;
+
+  public DefaultArtifactTypeLoader() {
+
+  }
+
+  public DefaultArtifactTypeLoader(Collection<ExtensionModel> extensionModels) {
+    this.extensionModels = extensionModels;
+  }
+
+  @Override
+  public void initialise() throws InitialisationException {
+    if (extensionModels == null) {
+      extensionModels = extensionManager.getExtensions();
+    }
```

Comment:
Shouldn't we throw an `InitializationException` if `extensionManager` is `null`? It could happen if the initialization is done without injection (and the no-args constructor was used).

Reply:
I will take the suggestion from above and add a nullCheck in the constructor.